rate_limiter/rate_limiter.go

```diff
@@ -0,0 +1,78 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package rate_limiter
+
+import (
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/armon/go-metrics"
+	"github.com/hashicorp/go-cleanhttp"
+	"golang.org/x/time/rate"
+)
+
+// CustomRoundTripper wraps http.RoundTripper to observe metrics and rate limit if necessary
+type CustomRoundTripper struct {
+	rateLimiter *rate.Limiter
+	source      string
+	rt          http.RoundTripper
+}
+
+func (irt *CustomRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
```

Comment:
Extremely nitpicky :grin: , feel free to ignore:

```suggestion
func (crt *CustomRoundTripper) RoundTrip(req *http.Request) (*http.Response, error) {
```